# Ejercicio 3

### Importación de librerías

In [19]:
import duckdb
import pandas as pd
from pathlib import Path
pd.set_option("display.max_columns", None)

### Rutas de archivos

In [2]:
RUTA_YELLOW = "../data/raw/yellow/2026/*.parquet"
RUTA_GREEN = "../data/raw/green/2026/*.parquet"

### Conexión de DuckDB

In [3]:
con = duckdb.connect()

### 3.1) Cantidad de archivos disponibles

In [4]:
archivos_yellow = list(Path('../data/raw/yellow/2026').glob("*.parquet"))
archivos_green = list(Path('../data/raw/green/2026').glob("*.parquet"))

print("Archivos de taxis amarillos:", len(archivos_yellow))
print("Archivos de taxis verdes:", len(archivos_green))
print("Total de archivos:", len(archivos_yellow) + len(archivos_green))

Archivos de taxis amarillos: 8
Archivos de taxis verdes: 8
Total de archivos: 16


### 3.2) Cantidad de registros disponibles

In [5]:
query = Path("../sql/ejercicio_3_2_a.sql").read_text()
registros_yellow = con.execute(query).fetchone()[0]

query = Path("../sql/ejercicio_3_2_b.sql").read_text()
registros_green = con.execute(query).fetchone()[0]

print("Registros de taxis amarillos:", registros_yellow)
print("Registros de taxis verdes:", registros_green)
print("Total de registros:", registros_yellow + registros_green)

Registros de taxis amarillos: 29703355
Registros de taxis verdes: 337114
Total de registros: 30040469


### 3.3) Columnas disponibles
### 3.4) Tipos de datos

In [6]:
query = Path("../sql/ejercicio_3_3_a.sql").read_text()
columnas_yellow = con.execute(query).df()

columnas_yellow

,column_name,column_type,null,key,default,extra
0,VendorID,INTEGER,YES,None,None,None
1,tpep_pickup_datetime,TIMESTAMP,YES,None,None,None
2,tpep_dropoff_datetime,TIMESTAMP,YES,None,None,None
3,passenger_count,BIGINT,YES,None,None,None
4,trip_distance,DOUBLE,YES,None,None,None
5,RatecodeID,BIGINT,YES,None,None,None
6,store_and_fwd_flag,VARCHAR,YES,None,None,None
7,PULocationID,INTEGER,YES,None,None,None
8,DOLocationID,INTEGER,YES,None,None,None
9,payment_type,BIGINT,YES,None,None,None


In [7]:
query = Path("../sql/ejercicio_3_3_b.sql").read_text()
columnas_green = con.execute(query).df()

columnas_green

,column_name,column_type,null,key,default,extra
0,VendorID,INTEGER,YES,None,None,None
1,lpep_pickup_datetime,TIMESTAMP,YES,None,None,None
2,lpep_dropoff_datetime,TIMESTAMP,YES,None,None,None
3,store_and_fwd_flag,VARCHAR,YES,None,None,None
4,RatecodeID,BIGINT,YES,None,None,None
5,PULocationID,INTEGER,YES,None,None,None
6,DOLocationID,INTEGER,YES,None,None,None
7,passenger_count,BIGINT,YES,None,None,None
8,trip_distance,DOUBLE,YES,None,None,None
9,fare_amount,DOUBLE,YES,None,None,None


3.3)  
Los taxis verdes tiene 2 columnas extras: **ehail_fee** y **trip_type**. Los taxis amarillos tienen una columna extra **Airport_fee**.  
Las dos variables de tipo Timestamp difieren en nombre en ambos conjuntos de datos, solo en la primera letra. Se homologará el nombre.  

3.4)  
- Los dos conjuntos tienen dos variables TIMESTAMP.
- Los dos conjuntos tienen las mismas 3 variables INTEGER.
- Los dos conjuntos tienen las mismas 3 variables BIGINT, adicional los taxis verdes tienen una variable más del tipo BIGINT **trip_type**.
- Los dos conjuntos tienen la misma variable VARCHAR.
- Los dos conjuntos tienen las mismas 10 variables DOUBLE, adicional los taxis verdes tienen una variable más del tipo DOUBLE **ehail_fee** y los taxis amarillos tienen **Airport_fee**.

### 3.5) Muestras de los conjuntos de datos

In [8]:
query = Path("../sql/ejercicio_3_5_a.sql").read_text()
muestra_yellow = con.execute(query).df()

muestra_yellow

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
0,2,2026-01-01 00:54:04,2026-01-01 00:59:37,1,0.97,1,N,239,238,1,7.2,1.00,0.5,3.66,0.0,1.0,15.86,2.5,0.0,0.00
1,1,2026-01-01 00:34:04,2026-01-01 00:39:47,0,0.90,1,N,163,162,2,7.9,4.25,0.5,0.00,0.0,1.0,13.65,2.5,0.0,0.75
2,1,2026-01-01 00:57:06,2026-01-01 01:05:59,0,1.40,1,N,43,237,1,10.7,4.25,0.5,2.50,0.0,1.0,18.95,2.5,0.0,0.75
3,2,2026-01-01 00:15:22,2026-01-01 00:58:10,4,5.58,1,N,142,209,1,38.7,1.00,0.5,11.11,0.0,1.0,55.56,2.5,0.0,0.75
4,2,2026-01-01 00:27:13,2026-01-01 00:40:43,0,2.16,1,N,88,144,1,13.5,1.00,0.5,3.85,0.0,1.0,23.10,2.5,0.0,0.75


In [9]:
query = Path("../sql/ejercicio_3_5_b.sql").read_text()
muestra_green = con.execute(query).df()

muestra_green

,VendorID,lpep_pickup_datetime,lpep_dropoff_datetime,store_and_fwd_flag,RatecodeID,PULocationID,DOLocationID,passenger_count,trip_distance,fare_amount,...,mta_tax,tip_amount,tolls_amount,ehail_fee,improvement_surcharge,total_amount,payment_type,trip_type,congestion_surcharge,cbd_congestion_fee
0,1,2026-01-01 00:27:58,2026-01-01 00:55:16,N,1,65,233,2,6.20,31.7,...,1.5,7.5,0.0,NaN,1.0,45.20,1,1,2.75,0.75
1,2,2026-01-01 00:44:33,2026-01-01 01:32:56,N,5,66,188,5,5.36,50.0,...,0.0,10.2,0.0,NaN,1.0,61.20,1,2,0.00,0.00
2,1,2026-01-01 00:23:45,2026-01-01 00:45:03,N,1,65,179,4,10.60,41.5,...,1.5,2.0,0.0,NaN,1.0,46.00,1,1,0.00,0.00
3,1,2026-01-01 00:44:33,2026-01-01 01:00:45,N,1,42,141,1,4.20,19.8,...,1.5,0.0,0.0,NaN,1.0,25.05,2,1,2.75,0.00
4,2,2026-01-01 00:46:04,2026-01-01 01:04:40,N,1,95,82,1,2.76,19.1,...,0.5,0.0,0.0,NaN,1.0,21.60,2,1,0.00,0.00


### 3.6) Posibles problemas de calidad de datos durante el análisis exploratorio
### 3.7) Consultas SQL
### 3.8) Documentación

##### Cantidad de nulos por columna

In [15]:
query = Path("../sql/ejercicio_3_6_a1.sql").read_text()
nulos_yellow = con.execute(query).df()

nulos_yellow

,VendorID_nulls,tpep_pickup_datetime_nulls,tpep_dropoff_datetime_nulls,store_and_fwd_flag_nulls,RatecodeID_nulls,PULocationID_nulls,DOLocationID_nulls,passenger_count_nulls,trip_distance_nulls,fare_amount_nulls,extra_nulls,mta_tax_nulls,tip_amount_nulls,tolls_amount_nulls,Airport_fee_nulls,improvement_surcharge_nulls,total_amount_nulls,payment_type_nulls,congestion_surcharge_nulls,cbd_congestion_fee_nulls
0,0,0,0,7716688,7716688,0,0,7716688,0,0,0,0,0,0,7716688,0,0,0,7716688,0


$\textbf{Documentación}$  

**Objetivo**  
Obtener la cantidad de valores nulos de cada columna.

**Consulta SQL:**  
```sql
SELECT
COUNT(*) FILTER (WHERE VendorID IS NULL) AS VendorID_nulls,
COUNT(*) FILTER (WHERE tpep_pickup_datetime IS NULL) AS tpep_pickup_datetime_nulls,
COUNT(*) FILTER (WHERE tpep_dropoff_datetime IS NULL) AS tpep_dropoff_datetime_nulls,
COUNT(*) FILTER (WHERE store_and_fwd_flag IS NULL) AS store_and_fwd_flag_nulls,
COUNT(*) FILTER (WHERE RatecodeID IS NULL) AS RatecodeID_nulls,
COUNT(*) FILTER (WHERE PULocationID IS NULL) AS PULocationID_nulls,
COUNT(*) FILTER (WHERE DOLocationID IS NULL) AS DOLocationID_nulls,
COUNT(*) FILTER (WHERE passenger_count IS NULL) AS passenger_count_nulls,
COUNT(*) FILTER (WHERE trip_distance IS NULL) AS trip_distance_nulls,
COUNT(*) FILTER (WHERE fare_amount IS NULL) AS fare_amount_nulls,
COUNT(*) FILTER (WHERE extra IS NULL) AS extra_nulls,
COUNT(*) FILTER (WHERE mta_tax IS NULL) AS mta_tax_nulls,
COUNT(*) FILTER (WHERE tip_amount IS NULL) AS tip_amount_nulls,
COUNT(*) FILTER (WHERE tolls_amount IS NULL) AS tolls_amount_nulls,
COUNT(*) FILTER (WHERE Airport_fee IS NULL) AS Airport_fee_nulls,
COUNT(*) FILTER (WHERE improvement_surcharge IS NULL) AS improvement_surcharge_nulls,
COUNT(*) FILTER (WHERE total_amount IS NULL) AS total_amount_nulls,
COUNT(*) FILTER (WHERE payment_type IS NULL) AS payment_type_nulls,
COUNT(*) FILTER (WHERE congestion_surcharge IS NULL) AS congestion_surcharge_nulls,
COUNT(*) FILTER (WHERE cbd_congestion_fee IS NULL) AS cbd_congestion_fee_nulls
FROM read_parquet('../data/raw/yellow/2026/*.parquet')
```

**Archivos usados**  
Se usaron los datos de taxis amarillos desde enero a septiembre de 2026.

**Resultados**  
Se tienen 7,716,688 valores nulos para las columnas store_and_fwd_flag, RatecodeID, passenger_count, Airport_fee y congestion_surcharge. Ninguna otra columna tiene faltantes. Ninguna otra columna tiene faltantes.

**Decisión**  
- store_and_fwd_flag: Sus nulos pueden deberse a la falla de conexión, por lo que podría reemplazarse con 'N' (véase el diccionario de datos en la carpeta docs).
- RatecodeID: No es relevante para el análisis, pero se puede cambiar por 99 (véase el diccionario de datos en la carpeta docs).
- passenger_count: Se evaluarán los mínimos de conteos y otras métricas, ya que estos pueden deberse a 0 pasajeros, lo que indicaría que estos viajes nunca sucedieron sino son errores. Si es el caso, las columnas se deberán eliminar.
- Airport_fee: Se puede deber a que es 0, ya que indica recogidas en aeropuertos específicos.
- congestion_surcharge: Puede deberse a que no existe congestión, por lo que podría ser reemplazado con 0.

In [20]:
query = Path("../sql/ejercicio_3_6_b1.sql").read_text()
nulos_green = con.execute(query).df()

nulos_green

,VendorID_nulls,lpep_pickup_datetime_nulls,lpep_dropoff_datetime_nulls,store_and_fwd_flag_nulls,RatecodeID_nulls,PULocationID_nulls,DOLocationID_nulls,passenger_count_nulls,trip_distance_nulls,fare_amount_nulls,extra_nulls,mta_tax_nulls,tip_amount_nulls,tolls_amount_nulls,ehail_fee_nulls,improvement_surcharge_nulls,total_amount_nulls,payment_type_nulls,trip_type_nulls,congestion_surcharge_nulls,cbd_congestion_fee_nulls
0,0,0,0,48775,48775,0,0,48775,0,0,0,0,0,0,337114,0,0,48775,48777,48775,0


$\textbf{Documentación}$  

**Objetivo**  
Obtener la cantidad de valores nulos de cada columna.

**Consulta SQL:**  
```sql
SELECT
COUNT(*) FILTER (WHERE VendorID IS NULL) AS VendorID_nulls,
COUNT(*) FILTER (WHERE lpep_pickup_datetime IS NULL) AS lpep_pickup_datetime_nulls,
COUNT(*) FILTER (WHERE lpep_dropoff_datetime IS NULL) AS lpep_dropoff_datetime_nulls,
COUNT(*) FILTER (WHERE store_and_fwd_flag IS NULL) AS store_and_fwd_flag_nulls,
COUNT(*) FILTER (WHERE RatecodeID IS NULL) AS RatecodeID_nulls,
COUNT(*) FILTER (WHERE PULocationID IS NULL) AS PULocationID_nulls,
COUNT(*) FILTER (WHERE DOLocationID IS NULL) AS DOLocationID_nulls,
COUNT(*) FILTER (WHERE passenger_count IS NULL) AS passenger_count_nulls,
COUNT(*) FILTER (WHERE trip_distance IS NULL) AS trip_distance_nulls,
COUNT(*) FILTER (WHERE fare_amount IS NULL) AS fare_amount_nulls,
COUNT(*) FILTER (WHERE extra IS NULL) AS extra_nulls,
COUNT(*) FILTER (WHERE mta_tax IS NULL) AS mta_tax_nulls,
COUNT(*) FILTER (WHERE tip_amount IS NULL) AS tip_amount_nulls,
COUNT(*) FILTER (WHERE tolls_amount IS NULL) AS tolls_amount_nulls,
COUNT(*) FILTER (WHERE ehail_fee IS NULL) AS ehail_fee_nulls,
COUNT(*) FILTER (WHERE improvement_surcharge IS NULL) AS improvement_surcharge_nulls,
COUNT(*) FILTER (WHERE total_amount IS NULL) AS total_amount_nulls,
COUNT(*) FILTER (WHERE payment_type IS NULL) AS payment_type_nulls,
COUNT(*) FILTER (WHERE trip_type IS NULL) AS trip_type_nulls,
COUNT(*) FILTER (WHERE congestion_surcharge IS NULL) AS congestion_surcharge_nulls,
COUNT(*) FILTER (WHERE cbd_congestion_fee IS NULL) AS cbd_congestion_fee_nulls
FROM read_parquet('../data/raw/green/2026/*.parquet')
```

**Archivos usados**  
Se usaron los datos de taxis verdes desde enero a septiembre de 2026.

**Resultados**  
Se tienen 48,775 valores nulos para las columnas store_and_fwd_flag, RatecodeID, passenger_count, payment_type, trip_type y congestion_surcharge. La columna ehail_feee tiene 33,7114 nulos. Ninguna otra columna tiene faltantes.

**Decisión**   
- store_and_fwd_flag: Sus nulos pueden deberse a la falla de conexión, por lo que podría reemplazarse con 'N' (véase el diccionario de datos en la carpeta docs).
- RatecodeID: No es relevante para el análisis, pero se puede cambiar por 99 (véase el diccionario de datos en la carpeta docs).
- payment_type: Podría reemplazarse con 5, ya que esto es desconocido (véase el diccionario de datos en la carpeta docs).
- trip_type: Podría reemplazarse con 1 o 2 (la moda), ya que estos son los únicos valores posibles (véase el diccionario de datos en la carpeta docs).
- passenger_count: Se evaluarán los mínimos de conteos y otras métricas, ya que estos pueden deberse a 0 pasajeros, lo que indicaría que estos viajes nunca sucedieron sino son errores. Si es el caso, las columnas se deberán eliminar.
- ehail_fee: No aparece en el diccionario, pero los nulos pueden deberse al ser una variable nueva, además todos son nulos, por lo que se eliminará esta columna.
- congestion_surcharge: Puede deberse a que no existe congestión, por lo que podría ser reemplazado con 0.

##### Distancias, monto y cantidad de pasajeros negativos y fechas inconsistentes

In [23]:
query = Path("../sql/ejercicio_3_6_a2.sql").read_text()
calidad_yellow = con.execute(query).df()

calidad_yellow

,total_amount_negativa,trip_distance_negativa,passenger_count_negativo,fechas_inconsistentes
0,161835,0,0,10


$\textbf{Documentación}$  

**Objetivo**  
Obtener la cantidad de datos incongruentes como distancias y montos negativos, y fechas inconsistentes entre inicio y fin del viaje.

**Consulta SQL:**  
```sql
SELECT
COUNT(*) FILTER (WHERE total_amount < 0) AS total_amount_negativa,
COUNT(*) FILTER (WHERE trip_distance < 0) AS trip_distance_negativa,
COUNT(*) FILTER (WHERE passenger_count < 0) AS passenger_count_negativo,
COUNT(*) FILTER (WHERE tpep_dropoff_datetime < tpep_pickup_datetime) AS fechas_inconsistentes
FROM read_parquet('../data/raw/yellow/2026/*.parquet')
```

**Archivos usados**  
Se usaron los datos de taxis amarillos desde enero a septiembre de 2026.

**Resultados**  
Se tienen 10 fechas inconsistentes y ninguna distancia o cantidad de pasajeros negativo. total_amunt tiene negativos.

**Decisión**  
- Las fechas inconsitentes se pueden eliminar al ser muy pocas.
- Las filas con total_amunt pueden quitarse al ser pocas.

In [24]:
query = Path("../sql/ejercicio_3_6_b2.sql").read_text()
calidad_green = con.execute(query).df()

calidad_green

,total_amount_negativa,trip_distance_negativa,passenger_count_negativo,fechas_inconsistentes
0,1023,0,0,5


$\textbf{Documentación}$  

**Objetivo**  
Obtener la cantidad de datos incongruentes como distancias y montos negativos, y fechas inconsistentes entre inicio y fin del viaje.

**Consulta SQL:**  
```sql
SELECT
COUNT(*) FILTER (WHERE trip_distance < 0) AS trip_distance_negativa,
COUNT(*) FILTER (WHERE passenger_count < 0) AS passenger_count_negativo,
COUNT(*) FILTER (WHERE lpep_dropoff_datetime < lpep_pickup_datetime) AS fechas_inconsistentes
FROM read_parquet('../data/raw/green/2026/*.parquet')
```

**Archivos usados**  
Se usaron los datos de taxis verdes desde enero a septiembre de 2026.

**Resultados**  
Se tienen 5 fechas inconsistentes y ninguna distancia o cantidad de pasajeros negativo. total_amunt tiene negativos.

**Decisión**
- Las fechas inconsitentes se pueden eliminar al ser muy pocas.
- Las filas con total_amunt pueden quitarse al ser pocas.

##### Mínimos, máximos, promedio y modas

In [33]:
query = Path("../sql/ejercicio_3_6_a3.sql").read_text()
estadisticas_yellow = con.execute(query).df()

estadisticas_yellow

,store_and_fwd_flag_mode,passenger_count_min,passenger_count_avg,passenger_count_max,Airport_fee_min,Airport_fee_avg,Airport_fee_max,congestion_surcharge_min,congestion_surcharge_avg,congestion_surcharge_max
0,N,0,1.249432,9,-2.0,0.16707,27.0,-2.5,2.216918,2.75


$\textbf{Documentación}$  

**Objetivo**  
Obtener estadísticas básicas de columnas con nulos.

**Consulta SQL:**  
```sql
SELECT
MODE(store_and_fwd_flag) AS store_and_fwd_flag_mode,
MIN(passenger_count) AS passenger_count_min,
AVG(passenger_count) AS passenger_count_avg,
MAX(passenger_count) AS passenger_count_max,
MIN(Airport_fee) AS Airport_fee_min,
AVG(Airport_fee) AS Airport_fee_avg,
MAX(Airport_fee) AS Airport_fee_max,
MIN(congestion_surcharge) AS congestion_surcharge_min,
AVG(congestion_surcharge) AS congestion_surcharge_avg,
MAX(congestion_surcharge) AS congestion_surcharge_max
FROM read_parquet('../data/raw/yellow/2026/*.parquet')
```

**Archivos usados**  
Se usaron los datos de taxis amarillos desde enero a septiembre de 2026.

**Resultados**  
Se tienen datos de pasajeros de 0, lo cual podría ser de viajes no realizados, lo cual podría indicarnos que estos viajes no los deberíamos de tomar en cuenta. Airport_fee_min puede deberse a una devolución. La moda de store_and_fwd_flag es 'N'.

**Decisión**  
- passenger_count: Los viajes con 0 pasajeros podrían eliminarse.
- Airport_fee_min: Se deberá revisar el significado de negativos.
- store_and_fwd_flag: Sus nulos se cambiarán por la moda.
- congestion_surcharge: Se deberá revisar el significado de negativos.

In [39]:
query = Path("../sql/ejercicio_3_6_b3.sql").read_text()
estadisticas_green = con.execute(query).df()

estadisticas_green

,store_and_fwd_flag_mode,trip_type_mode,payment_type_mode,passenger_count_min,passenger_count_avg,passenger_count_max,congestion_surcharge_min,congestion_surcharge_avg,congestion_surcharge_max
0,N,1,1,0,1.300646,9,-2.75,0.883127,2.75


$\textbf{Documentación}$  

**Objetivo**  
Obtener estadísticas básicas de columnas con nulos.

**Consulta SQL:**  
```sql
SELECT
MODE(store_and_fwd_flag) AS store_and_fwd_flag_mode,
MODE(trip_type) AS trip_type_mode,
MODE(payment_type) AS payment_type_mode,
MIN(passenger_count) AS passenger_count_min,
AVG(passenger_count) AS passenger_count_avg,
MAX(passenger_count) AS passenger_count_max,
MIN(ehail_fee) AS ehail_fee_min,
AVG(ehail_fee) AS ehail_fee_avg,
MAX(ehail_fee) AS ehail_fee_max,
MIN(congestion_surcharge) AS congestion_surcharge_min,
AVG(congestion_surcharge) AS congestion_surcharge_avg,
MAX(congestion_surcharge) AS congestion_surcharge_max
FROM read_parquet('../data/raw/green/2026/*.parquet')
```

**Archivos usados**  
Se usaron los datos de taxis verdes desde enero a septiembre de 2026.

**Resultados**  
Se tienen datos de pasajeros de 0, lo cual podría ser de viajes no realizados, lo cual podría indicarnos que estos viajes no los deberíamos de tomar en cuenta. La moda de store_and_fwd_flag es 'N'. la de trip_type es 1 y la de payment_type es 1.

**Decisión**  
- passenger_count: Los viajes con 0 pasajeros podrían eliminarse.
- store_and_fwd_flag: Sus nulos se cambiarán por la moda.
- trip_type: Sus nulos se cambiarán por la moda.
- payment_type: Sus nulos se cambiarán por la moda.
- congestion_surcharge: Se deberá revisar el significado de negativos.


##### Análisis de valores nulos y 0's en conteo de pasajeros con montos positivos 

In [52]:
query = Path("../sql/ejercicio_3_6_a4.sql").read_text()
estadisticas_pasajeros_invalidos_yellow = con.execute(query).df()

estadisticas_pasajeros_invalidos_yellow

,conteo,total_amount_min,total_amount_avg,total_amount_max,trip_distance_min,trip_distance_avg,trip_distance_max
0,7807001,0.5,32.495597,5000.0,0.0,11.675215,328522.2


$\textbf{Documentación}$  

**Objetivo**  
Obtener estadísticas básicas de total_amount y trip_distance para valores nulos y 0 de passenger_count y valores poditivos de total_amount.

**Consulta SQL:**  
```sql
SELECT
COUNT(*) AS conteo,
MIN(total_amount) AS total_amount_min,
AVG(total_amount) AS total_amount_avg,
MAX(total_amount) AS total_amount_max,
MIN(trip_distance) AS trip_distance_min,
AVG(trip_distance) AS trip_distance_avg,
MAX(trip_distance) AS trip_distance_max
FROM read_parquet('../data/raw/yellow/2026/*.parquet')
WHERE (passenger_count IS NULL
    OR passenger_count = 0)
    AND total_amount > 0
```

**Archivos usados**  
Se usaron los datos de taxis amarillos desde enero a septiembre de 2026.

**Resultados**  
Si se tienen valores congruentes de distancia y monto, la distancia 0 puede deberse a la cancelación de un viaje, pero que se haya cobrado una tarifa.

**Decisión**  
- passenger_count: Sus nulos y 0's se cambiarán por 1, ya que al menos 1 pasajero debió pedir el viaje.

In [53]:
query = Path("../sql/ejercicio_3_6_b4.sql").read_text()
estadisticas_pasajeros_invalidos_green = con.execute(query).df()

estadisticas_pasajeros_invalidos_green

,conteo,total_amount_min,total_amount_avg,total_amount_max,trip_distance_min,trip_distance_avg,trip_distance_max
0,52916,2.0,30.291746,287.37,0.0,70.412614,179830.92


$\textbf{Documentación}$  

**Objetivo**  
Obtener estadísticas básicas de total_amount y trip_distance para valores nulos y 0 de passenger_count y valores poditivos de total_amount.

**Consulta SQL:**  
```sql
SELECT
COUNT(*) AS conteo,
MIN(total_amount) AS total_amount_min,
AVG(total_amount) AS total_amount_avg,
MAX(total_amount) AS total_amount_max,
MIN(trip_distance) AS trip_distance_min,
AVG(trip_distance) AS trip_distance_avg,
MAX(trip_distance) AS trip_distance_max
FROM read_parquet('../data/raw/green/2026/*.parquet')
WHERE (passenger_count IS NULL
    OR passenger_count = 0)
    AND total_amount > 0
```

**Archivos usados**  
Se usaron los datos de taxis verdes desde enero a septiembre de 2026.

**Resultados**  
Si se tienen valores congruentes de distancia y monto, la distancia 0 puede deberse a la cancelación de un viaje, pero que se haya cobrado una tarifa.

**Decisión**  
- passenger_count: Sus nulos y 0's se cambiarán por 1, ya que al menos 1 pasajero debió pedir el viaje.

##### Valor máximo de total_amount para Airport_fee < 0 o congestion_surchaRge < 0

In [56]:
query = Path("../sql/ejercicio_3_6_a5.sql").read_text()
maximo_amount_yellow = con.execute(query).df()

maximo_amount_yellow

,max(total_amount)
0,-2.0


$\textbf{Documentación}$  

**Objetivo**  
Obtener el máximo de total_amount cuando Airport_fee < 0 o congestion_surchaRge < 0

**Consulta SQL:**  
```sql
SELECT MAX(total_amount)
FROM read_parquet('../data/raw/yellow/2026/*.parquet')
WHERE Airport_fee < 0
    OR congestion_surchaRge < 0
```

**Archivos usados**  
Se usaron los datos de taxis amarillos desde enero a septiembre de 2026.

**Resultados**  
El máximo es negativo.

**Decisión**  
Al eliminar los total_amount negativos se eliminarán los Airport_fee < 0 o congestion_surchaRge < 0.

In [57]:
query = Path("../sql/ejercicio_3_6_b5.sql").read_text()
maximo_amount_green = con.execute(query).df()

maximo_amount_green

,max(total_amount)
0,-4.25


$\textbf{Documentación}$ 

**Objetivo**  
Obtener el máximo de total_amount cuando congestion_surchaRge < 0

**Consulta SQL:**  
```sql
SELECT MAX(total_amount)
FROM read_parquet('../data/raw/yellow/2026/*.parquet')
WHERE congestion_surchaRge < 0
```

**Archivos usados**  
Se usaron los datos de taxis verdes desde enero a septiembre de 2026.

**Resultados**  
El máximo es negativo.

**Decisión**  
Al eliminar los total_amount negativos se eliminarán los congestion_surchaRge < 0.

##### Limpieza de datos

In [61]:
query = Path("../sql/ejercicio_3_6_a6.sql").read_text()
limpieza_yellow = con.execute(query).df()
limpieza_yellow.head()

,VendorID,pep_pickup_datetime,pep_dropoff_datetime,store_and_fwd_flag,RatecodeID,PULocationID,DOLocationID,passenger_count,trip_distance,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,congestion_surcharge,cbd_congestion_fee,total_amount,payment_type,Airport_fee
0,2,2026-04-22 08:49:55,2026-04-22 09:11:23,N,1,142,233,1,2.13,19.1,0.0,0.5,4.77,0.0,1.0,2.5,0.75,28.62,1,0.0
1,2,2026-04-22 08:18:25,2026-04-22 08:37:00,N,1,68,231,1,3.28,20.5,0.0,0.5,5.05,0.0,1.0,2.5,0.75,30.30,1,0.0
2,2,2026-04-22 08:05:41,2026-04-22 08:13:00,N,1,238,142,1,1.53,10.0,0.0,0.5,0.00,0.0,1.0,2.5,0.00,14.00,1,0.0
3,2,2026-04-22 08:47:04,2026-04-22 09:05:24,N,1,100,231,1,3.46,19.8,0.0,0.5,4.91,0.0,1.0,2.5,0.75,29.46,1,0.0
4,2,2026-04-22 08:51:39,2026-04-22 09:07:03,N,1,162,48,1,1.23,14.2,0.0,0.5,3.79,0.0,1.0,2.5,0.75,22.74,1,0.0


$\textbf{Documentación}$  

**Objetivo**  
Limpiar los datos para usarlos en análisis posteriores.

**Consulta SQL:**  
```sql
WITH datos AS (
    SELECT *
    FROM read_parquet('../data/raw/yellow/2026/*.parquet')
),
modas AS (
    SELECT mode(store_and_fwd_flag) AS moda_store_and_fwd_flag
    FROM datos
)
SELECT
    VendorID,
    tpep_pickup_datetime AS pep_pickup_datetime,
    tpep_dropoff_datetime AS pep_dropoff_datetime,
    COALESCE(store_and_fwd_flag, moda_store_and_fwd_flag) AS store_and_fwd_flag,
    COALESCE(RatecodeID, 99) AS RatecodeID,
    PULocationID,
    DOLocationID,
    CASE
        WHEN passenger_count IS NULL OR passenger_count = 0 THEN 1
        ELSE passenger_count
    END AS passenger_count,
    trip_distance,
    fare_amount,
    extra,
    mta_tax,
    tip_amount,
    tolls_amount,
    improvement_surcharge,
    COALESCE(congestion_surcharge, 0) AS congestion_surcharge,
    cbd_congestion_fee,
    total_amount,
    payment_type,
    COALESCE(Airport_fee, 0) AS Airport_fee
FROM datos
CROSS JOIN modas
WHERE
    total_amount >= 0
    AND  tpep_dropoff_datetime >= tpep_pickup_datetime
```

**Archivos usados**  
Se usaron los datos de taxis amarillos desde enero a septiembre de 2026.

**Resultados**  
Se tienen datos limpios.

**Decisión**  
1.	vendor_ID: No se harán cambios.
2.	tpep_pickup_datetime: Se cambia el nombre a pep_pickup_datetime. Se eliminan las fechas inconsistentes.
3.	tpep_dropoff_datetime: Se cambia el nombre a pep_dropoff_datetime. Se eliminan las fechas inconsistentes.
4.	store_and_fwd_flag: Los nulos se cambiarán por la moda.
5.	RatecodeID: Los nulos se cambian por 99.
6.	PULocationID: No se harán cambios.
7.	DOLocationID: No se harán cambios.
8.	passenger_count: Los nulos y 0’s se cambiarán por 1.
9.	trip_distance: No se harán cambios.
10.	fare_amount: No se harán cambios.
11.	extra: No se harán cambios.
12.	mta_tax: No se harán cambios.
13.	tip_amount: No se harán cambios.
14.	tolls_amount: No se harán cambios.
15.	improvement_surcharge: No se harán cambios.
16.	congestion_surcharge: Los nulos se cambiarán por 0.
17.	cbd_congestion_fee: No se harán cambios.
18.	total_amount: Se eleminan datos negativos.
19.	payment_type: No se harán cambios.
20.	Airport_fee: Los nulos serán reemplazados con 0.


In [62]:
query = Path("../sql/ejercicio_3_6_b6.sql").read_text()
limpieza_green = con.execute(query).df()
limpieza_green.head()

,VendorID,pep_pickup_datetime,pep_dropoff_datetime,store_and_fwd_flag,RatecodeID,PULocationID,DOLocationID,passenger_count,trip_distance,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,congestion_surcharge,cbd_congestion_fee,total_amount,payment_type,trip_type
0,1,2026-01-01 00:27:58,2026-01-01 00:55:16,N,1,65,233,2,6.20,31.7,4.50,1.5,7.5,0.0,1.0,2.75,0.75,45.20,1,1
1,2,2026-01-01 00:44:33,2026-01-01 01:32:56,N,5,66,188,5,5.36,50.0,0.00,0.0,10.2,0.0,1.0,0.00,0.00,61.20,1,2
2,1,2026-01-01 00:23:45,2026-01-01 00:45:03,N,1,65,179,4,10.60,41.5,1.00,1.5,2.0,0.0,1.0,0.00,0.00,46.00,1,1
3,1,2026-01-01 00:44:33,2026-01-01 01:00:45,N,1,42,141,1,4.20,19.8,3.75,1.5,0.0,0.0,1.0,2.75,0.00,25.05,2,1
4,2,2026-01-01 00:46:04,2026-01-01 01:04:40,N,1,95,82,1,2.76,19.1,1.00,0.5,0.0,0.0,1.0,0.00,0.00,21.60,2,1


$\textbf{Documentación}$  

**Objetivo**  
Limpiar los datos para usarlos en análisis posteriores.

**Consulta SQL:**  
```sql
WITH datos AS (
    SELECT *
    FROM read_parquet('../data/raw/green/2026/*.parquet')
),
modas AS (
    SELECT
        mode(store_and_fwd_flag) AS moda_store_and_fwd_flag,
        mode(payment_type) AS moda_payment_type,
        mode(trip_type) AS moda_trip_type
    FROM datos
)
SELECT
    VendorID,
    lpep_pickup_datetime AS pep_pickup_datetime,
    lpep_dropoff_datetime AS pep_dropoff_datetime,
    COALESCE(store_and_fwd_flag, moda_store_and_fwd_flag) AS store_and_fwd_flag,
    COALESCE(RatecodeID, 99) AS RatecodeID,
    PULocationID,
    DOLocationID,
    CASE
        WHEN passenger_count IS NULL OR passenger_count = 0 THEN 1
        ELSE passenger_count
    END AS passenger_count,
    trip_distance,
    fare_amount,
    extra,
    mta_tax,
    tip_amount,
    tolls_amount,
    improvement_surcharge,
    COALESCE(congestion_surcharge, 0) AS congestion_surcharge,
    cbd_congestion_fee,
    total_amount,
    COALESCE(payment_type, moda_payment_type) AS payment_type,
    COALESCE(trip_type, moda_trip_type) AS trip_type
FROM datos
CROSS JOIN modas
WHERE
    total_amount >= 0
    AND lpep_dropoff_datetime >= lpep_pickup_datetime
```

**Archivos usados**  
Se usaron los datos de taxis verdes desde enero a septiembre de 2026.

**Resultados**  
Se tienen datos limpios.

**Decisión**  
1.	vendor_ID: No se harán cambios.
2.	tpep_pickup_datetime: Se cambia el nombre a pep_pickup_datetime. Se eliminan las fechas inconsistentes.
3.	tpep_dropoff_datetime: Se cambia el nombre a pep_dropoff_datetime. Se eliminan las fechas inconsistentes.
4.	store_and_fwd_flag: Los nulos se cambiarán por la moda.
5.	RatecodeID: Los nulos se cambian por 99.
6.	PULocationID: No se harán cambios.
7.	DOLocationID: No se harán cambios.
8.	passenger_count: Los nulos y 0’s se cambiarán por 1.
9.	trip_distance: No se harán cambios.
10.	fare_amount: No se harán cambios.
11.	extra: No se harán cambios.
12.	mta_tax: No se harán cambios.
13.	tip_amount: No se harán cambios.
14.	tolls_amount: No se harán cambios.
15.	improvement_surcharge: No se harán cambios.
16.	congestion_surcharge: Los nulos se cambiarán por 0.
17.	cbd_congestion_fee: No se harán cambios.
18.	total_amount: Se eleminan datos negativos.
19.	payment_type: Los nulos se cambiarán por la moda.
20.	trip_type: Los nulos se cambiarán por la moda.
21.	ehail_fee: Se eliminará esta columna.

### Guardando datos

In [63]:
limpieza_yellow.to_parquet("../data/processed/yellow_2026.parquet", index=False)
limpieza_green.to_parquet("../data/processed/green_2026.parquet", index=False)